// kaiser/lib/index.js
"use strict";

require("earlgrey-runtime/5");var tmp$0 = undefined;var $targ$11 = undefined;var accum$5 = undefined;var $targ$12 = undefined;var accum$6 = undefined;var accum$7 = undefined;var accum$8 = undefined;var accum$9 = undefined;var accum$10 = undefined;var accum$11 = undefined;var accum$12 = undefined;var m$27 = undefined;var $targ$19 = undefined;var $targ$20 = undefined;var $targ$21 = undefined;var $targ$22 = undefined;var $targ$23 = undefined;var $targ$24 = undefined;var $targ$25 = undefined;var $targ$26 = undefined;var $targ$27 = undefined;var $targ$28 = undefined;var $targ$29 = undefined;var $targ$30 = undefined;var $targ$31 = undefined;var $targ$32 = undefined;var TYPEID$0 = undefined;var registry$0 = undefined;var object__minus__name$0 = undefined;var make__minus__type__minus__id$0 = undefined;var register$0 = undefined;var register__minus__all$0 = undefined;var object__minus__serializer$0 = undefined;var singleton__minus__serializer$0 = undefined;var register__minus__singleton$0 = undefined;var register__minus__singletons$0 = undefined;var register__minus__function$0 = undefined;var register__minus__functions$0 = undefined;var Serializer$0 = undefined;var default__minus__serializer$0 = undefined;var serialize$0 = undefined;var deserialize$0 = undefined;var shallow__minus__copy$0 = undefined;var deep__minus__copy$0 = undefined;var cmds$0 = undefined;var $targ$0 = undefined;if ((tmp$0 = send(global, "@@KAISER", true), tmp$0 === undefined || tmp$0 === null ? tmp$0 : tmp$0.loaded)) {
  return global["@@KAISER"];
}TYPEID$0 = Symbol("TYPEID");registry$0 = {};object__minus__name$0 = function objectName() {
  var tmp$1 = undefined;var tmp$3 = undefined;var tmp$2 = undefined;var name$0 = undefined;var m$1$0 = undefined;var object$0 = undefined;var req$0 = undefined;var t0$0 = undefined;var m$0$0 = undefined;m$0$0 = arguments;t0$0 = m$0$0.length;if (t0$0 >= 1 && t0$0 <= 2) {
    object$0 = m$0$0[0];if (1 >= t0$0) {
      req$0 = false;
    } else {
      req$0 = m$0$0[1];
    }m$1$0 = (tmp$1 = object$0, tmp$1 === undefined || tmp$1 === null ? tmp$1 : tmp$1.name) || (tmp$2 = (tmp$3 = object$0, tmp$3 === undefined || tmp$3 === null ? tmp$3 : tmp$3.constructor), tmp$2 === undefined || tmp$2 === null ? tmp$2 : tmp$2.name);if (!m$1$0 && req$0) {
      throw ErrorFactory(["kaiser", "nameRequired"]).create("The object or function to register must have a name field or its constructor must have a name.");
    } else {
      if (!m$1$0) {
        return "";
      } else {
        name$0 = m$1$0;return name$0;
      }
    }
  } else {
    return ___match_error(m$0$0, "{object, req = false}");
  }
};make__minus__type__minus__id$0 = function makeTypeId(object$1, temp$0$0) {
  var t0$1 = undefined;var name__minus__variant$0 = undefined;var dv$0 = undefined;var data__minus__version$0 = undefined;var m$2$0 = undefined;var version$0 = undefined;var package$1 = undefined;var use__minus__version$0 = undefined;var type__minus__id$2 = undefined;var type__minus__id$1 = undefined;var ph$1$0 = undefined;var type__minus__id$0 = undefined;var variant$0 = undefined;var $$107$0 = undefined;var $$108$0 = undefined;var t0$2 = undefined;var policy$0 = undefined;var ph$0$0 = undefined;t0$1 = temp$0$0;policy$0 = t0$1;ph$0$0 = t0$1;t0$2 = ph$0$0;if (($$108$0 = ___hasprop(t0$2, "typeId")) && (type__minus__id$0 = t0$2.typeId, ___hasprop(t0$2, "variant"))) {
    variant$0 = t0$2.variant;return type__minus__id$0 + "/" + JSON.stringify(variant$0);
  } else {
    if ($$108$0 && (type__minus__id$1 = t0$2.typeId, ___hasprop(t0$2, "nameVariant"))) {
      ph$1$0 = t0$2.nameVariant;if (ph$1$0 === true) {
        return type__minus__id$1 + "/" + object__minus__name$0(object$1);
      } else {
        name__minus__variant$0 = ph$1$0;return type__minus__id$1 + "/" + name__minus__variant$0;
      }
    } else {
      if ($$108$0) {
        type__minus__id$2 = t0$2.typeId;return type__minus__id$2;
      } else {
        if (___hasprop(t0$2, "package")) {
          package$1 = t0$2["package"];if (___hasprop(t0$2, "useVersion")) {
            use__minus__version$0 = t0$2.useVersion;
          } else {
            use__minus__version$0 = null;
          }m$2$0 = use__minus__version$0;if (!m$2$0) {
            version$0 = "";
          } else {
            if ((dv$0 = package$1.dataVersion, dv$0)) {
              data__minus__version$0 = ENode([], {}, ["d", dv$0]).toString();
            } else {
              data__minus__version$0 = package$1.version;
            }if (m$2$0 === "major") {
              version$0 = ENode([], {}, ["@", data__minus__version$0.split(".")[0]]).toString();
            } else {
              if (m$2$0 === "minor") {
                version$0 = ENode([], {}, ["@", send(data__minus__version$0.split("."), range(0, 2 - 1)).join(".")]).toString();
              } else {
                if (m$2$0 === "patch") {
                  version$0 = ENode([], {}, ["@", data__minus__version$0]).toString();
                } else {
                  version$0 = ___match_error(m$2$0, ".patch");
                }
              }
            }
          }return ENode([], {}, ["npm:", package$1.name, version$0, "/", object__minus__name$0(object$1)]).toString();
        } else {
          throw ErrorFactory(["kaiser", "register", "id"]).create("No typeId provided.");
        }
      }
    }
  }
};register$0 = function register(object$2, _policy$0) {
  var t0$3 = undefined;var bridge$$215$0 = undefined;var bridge$$218$0 = undefined;var m$3$0 = undefined;var policy$1 = undefined;var $targ$1 = undefined;var type__minus__id$3 = undefined;var prev$0 = undefined;var $targ$2 = undefined;var $targ$3 = undefined;policy$1 = __amp__(object__minus__serializer$0(object$2), _policy$0);if (policy$1.target !== "self" && object$2.prototype && !policy$1.ignorePrototype) {
    return register$0(object$2.prototype, _policy$0);
  }$targ$1 = make__minus__type__minus__id$0(object$2, policy$1);t0$3 = $targ$1;policy$1.typeId = t0$3;type__minus__id$3 = t0$3;if (Object.isFrozen(object$2)) {
    throw ErrorFactory(["kaiser", "frozen"]).create("kaiser cannot register frozen objects as serializable types; " + "either register them before freezing them, or register their prototypes.");
  }prev$0 = send(registry$0, type__minus__id$3);if (prev$0 && prev$0 !== policy$1) {
    m$3$0 = prev$0.onDuplicate;bridge$$215$0 = m$3$0;if (bridge$$215$0 === "error" || policy$1.onDuplicate === "error") {
      throw ErrorFactory(["kaiser", "duplicateId"]).create(ENode([], {}, ["Duplicate typeId: \"", type__minus__id$3, "\". You can control this error"]).toString() + ENode([], {}, [" with the onDuplicate option to kaiser.register"]).toString() + ENode([], {}, [" (onDuplicate: \"error\" (default), \"(warn-)keep\" or \"(warn-)replace\")."]).toString());
    } else {
      if (m$3$0 === "warnKeep") {
        console.warn(ENode([], {}, ["WARNING: Duplicate typeId: \"", type__minus__id$3, "\". Keeping previous."]).toString());return false;
      } else {
        bridge$$218$0 = m$3$0;if (bridge$$218$0 === "warn" || bridge$$218$0 === "warnReplace") {
          console.warn(ENode([], {}, ["WARNING: Duplicate typeId: \"", type__minus__id$3, "\". Replacing."]).toString());
        } else {
          if (m$3$0 === "keep") {
            return false;
          } else {
            if (m$3$0 === "replace") {
              undefined;
            } else {
              ___match_error(m$3$0, ".replace");
            }
          }
        }
      }
    }
  }$targ$2 = policy$1;registry$0[type__minus__id$3] = $targ$2;$targ$3 = type__minus__id$3;object$2[TYPEID$0] = $targ$3;return true;
};register__minus__all$0 = function registerAll(objects$0, policy$2) {
  var m$4 = undefined;var acc$0 = undefined;var temp$1 = undefined;acc$0 = [];temp$1 = null;m$4 = null;$1: for (var _iterator = objects$0[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
    m$4 = _step.value;
    var object$3 = undefined;object$3 = m$4;temp$1 = register$0(object$3, __amp__(policy$2, { nameVariant: object__minus__name$0(object$3) }));acc$0.push(temp$1);
  }return acc$0;
};object__minus__serializer$0 = function objectSerializer(proto$0) {
  var accum$0 = undefined;var accum$1 = undefined;return __amp____colon__({ onDuplicate: "error", target: "direct", ignorePrototype: false }, __amp____colon__((accum$0 = {}, accum$0.serialize = function serialize(x$0) {
    return object(items(x$0));
  }, accum$0), (accum$1 = {}, accum$1.deserialize = function deserialize(x$1) {
    var m$5 = undefined;var rval$0 = undefined;rval$0 = Object.create(proto$0);m$5 = null;$2: for (var _iterator = items(x$1)[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
      m$5 = _step.value;
      var $targ$4 = undefined;var k$0 = undefined;var v$0 = undefined;var t0$4 = undefined;var t1$0 = undefined;t0$4 = m$5;if (Array.isArray(t0$4) && (t1$0 = t0$4.length, t1$0 === 2)) {
        k$0 = t0$4[0];v$0 = t0$4[1];$targ$4 = v$0;rval$0[k$0] = $targ$4;void 0;
      } else {
        ___match_error(m$5);
      }
    }return rval$0;
  }, accum$1)));
};singleton__minus__serializer$0 = function singletonSerializer(object$4) {
  var accum$2 = undefined;var accum$3 = undefined;return __amp____colon__({ target: "self" }, __amp____colon__((accum$2 = {}, accum$2.serialize = function serialize(temp$2$0) {
    return null;
  }, accum$2), (accum$3 = {}, accum$3.deserialize = function deserialize(temp$3$0) {
    return object$4;
  }, accum$3)));
};register__minus__singleton$0 = function registerSingleton() {
  var policy$3 = undefined;var type__minus__id$4 = undefined;var t0$6 = undefined;var object$5 = undefined;var ph$3$0 = undefined;var t0$5 = undefined;var m$6$0 = undefined;m$6$0 = arguments;t0$5 = m$6$0.length;if (t0$5 >= 1 && t0$5 <= 2) {
    object$5 = m$6$0[0];if (1 >= t0$5) {
      ph$3$0 = null;
    } else {
      ph$3$0 = m$6$0[1];
    }if (ph$3$0 === null) {
      return function (o$0) {
        return register__minus__singleton$0(o$0, object$5);
      };
    } else {
      t0$6 = ph$3$0;if (typeof t0$6 === "string") {
        type__minus__id$4 = t0$6;register$0(object$5, __amp__(singleton__minus__serializer$0(object$5), { typeId: type__minus__id$4 }));return object$5;
      } else {
        policy$3 = ph$3$0;register$0(object$5, __amp__(singleton__minus__serializer$0(object$5), policy$3));return object$5;
      }
    }
  } else {
    return ___match_error(m$6$0, "{object, match = null}");
  }
};register__minus__singletons$0 = function registerSingletons(objects$1, _policy$1) {
  var type__minus__id$5 = undefined;var m$7$0 = undefined;var m$8 = undefined;var acc$1 = undefined;var temp$4 = undefined;var policy$4 = undefined;m$7$0 = _policy$1;if (typeof m$7$0 === "string") {
    type__minus__id$5 = m$7$0;policy$4 = { typeId: type__minus__id$5 };
  } else {
    policy$4 = _policy$1;
  }acc$1 = [];temp$4 = null;m$8 = null;$3: for (var _iterator = objects$1[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
    m$8 = _step.value;
    var object$6 = undefined;object$6 = m$8;temp$4 = register$0(object$6, __amp__(singleton__minus__serializer$0(object$6), __amp__(policy$4, { nameVariant: object__minus__name$0(object$6) })));acc$1.push(temp$4);
  }return acc$1;
};register__minus__function$0 = register__minus__singleton$0;register__minus__functions$0 = register__minus__singletons$0;Serializer$0 = function Serializer() {
  var $targ$5 = undefined;var $targ$7 = undefined;var accum$4 = undefined;var t0$8 = undefined;var acquire$0 = undefined;var $targ$6 = undefined;var wl$0 = undefined;var white__minus__list$0 = undefined;var ph$5$0 = undefined;var t0$7 = undefined;var t1$1 = undefined;var m$9$0 = undefined;var __at___$0 = undefined;if (!getChecker(Serializer$0)(this)) {
    __at___$0 = Object.create(Serializer$0.prototype);
  } else {
    __at___$0 = this;
  }m$9$0 = arguments;t0$7 = m$9$0.length;if (t0$7 >= 0 && t0$7 <= 1) {
    if (0 >= t0$7) {
      t1$1 = null;
    } else {
      t1$1 = m$9$0[0];
    }white__minus__list$0 = t1$1;ph$5$0 = t1$1;if (ph$5$0 === null) {
      $targ$5 = null;__at___$0.whiteList = $targ$5;void 0;
    } else {
      $targ$6 = __amp____colon__({ native: true }, ($targ$7 = true, accum$4 = {}, accum$4["builtin:Object"] = $targ$7, accum$4));t0$8 = $targ$6;__at___$0.whiteList = t0$8;wl$0 = t0$8;acquire$0 = function acquire(temp$5$0) {
        var t0$9 = undefined;var m$10 = undefined;var acc$2 = undefined;var temp$6 = undefined;var $targ$8 = undefined;var $targ$9 = undefined;var $targ$10 = undefined;var tid$1 = undefined;var tid$0 = undefined;var $$532$0 = undefined;var t0$10 = undefined;var t1$2 = undefined;var t2$0 = undefined;var w$0 = undefined;var ph$6$0 = undefined;t0$9 = temp$5$0;w$0 = t0$9;ph$6$0 = t0$9;if (Array.isArray(ph$6$0)) {
          acc$2 = [];temp$6 = null;m$10 = null;$4: for (var _iterator = w$0[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
            m$10 = _step.value;
            var entry$0 = undefined;entry$0 = m$10;temp$6 = acquire$0(entry$0);acc$2.push(temp$6);
          }return acc$2;
        } else {
          if (typeof ph$6$0 === "string") {
            $targ$8 = true;wl$0[w$0] = $targ$8;return void 0;
          } else {
            t0$10 = ph$6$0;if (___hasprop(t0$10, "prototype") && (t1$2 = t0$10.prototype, t2$0 = TYPEID$0, ___hasprop(t1$2, t2$0))) {
              tid$0 = send(t1$2, t2$0);$targ$9 = true;wl$0[tid$0] = $targ$9;return void 0;
            } else {
              t1$2 = TYPEID$0;if (___hasprop(t0$10, t1$2)) {
                tid$1 = send(t0$10, t1$2);$targ$10 = true;wl$0[tid$1] = $targ$10;return void 0;
              } else {
                return ___match_error(ph$6$0, "{^TYPEID => tid}");
              }
            }
          }
        }
      };acquire$0(white__minus__list$0 || []);
    }
  } else {
    ___match_error(m$9$0, "{match white-list = null}");
  }return __at___$0;
};Serializer$0.prototype.toObject = function () {
  var bridge$$628$0 = undefined;var bridge$$627$0 = undefined;var bridge$$626$0 = undefined;var m$12 = undefined;var acc$3 = undefined;var temp$7 = undefined;var m$13 = undefined;var acc$4 = undefined;var temp$8 = undefined;var p$0 = undefined;var t0$12 = undefined;var policy$5 = undefined;var ph$9$0 = undefined;var m$14$0 = undefined;var bridge$$793$0 = undefined;var bridge$$792$0 = undefined;var bridge$$791$0 = undefined;var m$16 = undefined;var acc$5 = undefined;var temp$9 = undefined;var m$17 = undefined;var acc$6 = undefined;var temp$10 = undefined;var bridge$$790$0 = undefined;var m$15$0 = undefined;var tid$2 = undefined;var ser0$0 = undefined;var ser1$0 = undefined;var bridge$$625$0 = undefined;var obj$0 = undefined;var ph$8$0 = undefined;var recursive$0 = undefined;var t0$11 = undefined;var t1$3 = undefined;var m$11$0 = undefined;var __at___$1 = undefined;var self$0 = undefined;__at___$1 = this;self$0 = this;m$11$0 = arguments;t0$11 = m$11$0.length;if (t0$11 >= 1 && t0$11 <= 2) {
    t1$3 = m$11$0[0];obj$0 = t1$3;ph$8$0 = t1$3;if (1 >= t0$11) {
      recursive$0 = true;
    } else {
      recursive$0 = m$11$0[1];
    }bridge$$625$0 = ph$8$0;if ((bridge$$626$0 = bridge$$625$0, (bridge$$627$0 = bridge$$626$0, (bridge$$628$0 = bridge$$627$0, bridge$$628$0 === null || bridge$$628$0 === void 0) || typeof bridge$$627$0 === "boolean") || typeof bridge$$626$0 === "number") || typeof bridge$$625$0 === "string") {
      return { typeId: ENode([], {}, ["native"]).toString(), data: obj$0 };
    } else {
      if (Array.isArray(ph$8$0) && recursive$0) {
        acc$3 = [];temp$7 = null;m$12 = null;$5: for (var _iterator = obj$0[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
          m$12 = _step.value;
          var x$2 = undefined;x$2 = m$12;temp$7 = __at___$1.toObject(x$2);acc$3.push(temp$7);
        }return acc$3;
      } else {
        if (Array.isArray(ph$8$0)) {
          acc$4 = [];temp$8 = null;m$13 = null;$6: for (var _iterator2 = obj$0[Symbol.iterator](), _step2; !(_step2 = _iterator2.next()).done;) {
            m$13 = _step2.value;
            var x$3 = undefined;x$3 = m$13;temp$8 = x$3;acc$4.push(temp$8);
          }return acc$4;
        } else {
          tid$2 = send(obj$0, TYPEID$0);if (!tid$2) {
            throw ErrorFactory(["kaiser", "serialize"]).create(ENode([], {}, ["Object has no typeId and cannot be serialized: ", obj$0]).toString());
          }if (__at___$1.whiteList && !send(send(__at___$1, "whiteList", true), tid$2)) {
            throw ErrorFactory(["kaiser", "serialize"]).create(ENode([], {}, ["Object with type id \"", tid$2, "\" is not white-listed for serialization: ", obj$0]).toString());
          }m$14$0 = send(registry$0, tid$2);if (m$14$0 === void 0) {
            throw ErrorFactory(["kaiser", "serialize"]).create(ENode([], {}, ["Cannot serialize typeId: \"", tid$2, "\" for object ", obj$0]).toString());
          } else {
            policy$5 = m$14$0;if (___hasprop(m$14$0, "target")) {
              ph$9$0 = m$14$0.target;if (ph$9$0 === "self" && !Object.hasOwnProperty.call(obj$0, TYPEID$0)) {
                throw ErrorFactory(["kaiser", "serialize"]).create(ENode([], {}, ["Object has no typeId and cannot be serialized: ", obj$0, " [1]"]).toString());
              } else {
                t0$12 = ph$9$0;if (t0$12 === "direct" && (p$0 = Object.getPrototypeOf(obj$0), !p$0.hasOwnProperty.call(p$0, TYPEID$0))) {
                  throw ErrorFactory(["kaiser", "serialize"]).create(ENode([], {}, ["Object has no typeId and cannot be serialized: ", obj$0, " [2]"]).toString());
                } else {
                  ser0$0 = policy$5.serialize(obj$0);
                }
              }
            } else {
              ser0$0 = ___match_error(m$14$0, "policy and {target => match}");
            }
          }m$15$0 = ser0$0;if (!recursive$0) {
            ser1$0 = ser0$0;
          } else {
            bridge$$790$0 = m$15$0;if ((bridge$$791$0 = bridge$$790$0, (bridge$$792$0 = bridge$$791$0, (bridge$$793$0 = bridge$$792$0, bridge$$793$0 === null || bridge$$793$0 === void 0) || typeof bridge$$792$0 === "string") || typeof bridge$$791$0 === "number") || typeof bridge$$790$0 === "boolean") {
              ser1$0 = ser0$0;
            } else {
              if (Array.isArray(m$15$0)) {
                acc$5 = [];temp$9 = null;m$16 = null;$7: for (var _iterator3 = ser0$0[Symbol.iterator](), _step3; !(_step3 = _iterator3.next()).done;) {
                  m$16 = _step3.value;
                  var x$4 = undefined;x$4 = m$16;temp$9 = __at___$1.toObject(x$4);acc$5.push(temp$9);
                }ser1$0 = acc$5;
              } else {
                if (getChecker(Object)(m$15$0)) {
                  ser1$0 = object((acc$6 = [], temp$10 = null, m$17 = null, (function () {
                    $8: for (var _iterator4 = items(ser0$0)[Symbol.iterator](), _step4; !(_step4 = _iterator4.next()).done;) {
                      m$17 = _step4.value;
                      var k$1 = undefined;var v$1 = undefined;var t0$13 = undefined;var t1$4 = undefined;t0$13 = m$17;if (Array.isArray(t0$13) && (t1$4 = t0$13.length, t1$4 === 2)) {
                        k$1 = t0$13[0];v$1 = t0$13[1];temp$10 = [k$1, __at___$1.toObject(v$1)];acc$6.push(temp$10);
                      } else {
                        ___match_error(m$17);
                      }
                    }
                  })(), acc$6));
                } else {
                  ser1$0 = ___match_error(m$15$0, "Object? ");
                }
              }
            }
          }return { typeId: tid$2, data: ser1$0 };
        }
      }
    }
  } else {
    return ___match_error(m$11$0, "{match obj, recursive = true}");
  }
};Serializer$0.prototype.fromObject = function () {
  var m$19 = undefined;var acc$7 = undefined;var temp$11 = undefined;var m$20 = undefined;var acc$8 = undefined;var temp$12 = undefined;var data$0 = undefined;var bridge$$1037$0 = undefined;var bridge$$1036$0 = undefined;var bridge$$1035$0 = undefined;var m$24 = undefined;var acc$9 = undefined;var temp$13 = undefined;var m$25 = undefined;var acc$10 = undefined;var temp$14 = undefined;var bridge$$1034$0 = undefined;var m$23$0 = undefined;var policy$6 = undefined;var m$22$0 = undefined;var deser$0 = undefined;var m$21$0 = undefined;var tid$3 = undefined;var obj$1 = undefined;var ph$11$0 = undefined;var recursive$1 = undefined;var t0$14 = undefined;var t1$5 = undefined;var m$18$0 = undefined;var __at___$2 = undefined;var self$1 = undefined;__at___$2 = this;self$1 = this;m$18$0 = arguments;t0$14 = m$18$0.length;if (t0$14 >= 1 && t0$14 <= 2) {
    t1$5 = m$18$0[0];obj$1 = t1$5;ph$11$0 = t1$5;if (1 >= t0$14) {
      recursive$1 = true;
    } else {
      recursive$1 = m$18$0[1];
    }if (Array.isArray(ph$11$0) && recursive$1) {
      acc$7 = [];temp$11 = null;m$19 = null;$9: for (var _iterator = obj$1[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
        m$19 = _step.value;
        var x$5 = undefined;x$5 = m$19;temp$11 = __at___$2.fromObject(x$5);acc$7.push(temp$11);
      }return acc$7;
    } else {
      if (Array.isArray(ph$11$0)) {
        acc$8 = [];temp$12 = null;m$20 = null;$10: for (var _iterator2 = obj$1[Symbol.iterator](), _step2; !(_step2 = _iterator2.next()).done;) {
          m$20 = _step2.value;
          var x$6 = undefined;x$6 = m$20;temp$12 = x$6;acc$8.push(temp$12);
        }return acc$8;
      } else {
        tid$3 = obj$1.typeId;if (__at___$2.whiteList && !send(send(__at___$2, "whiteList", true), tid$3)) {
          throw ErrorFactory(["kaiser", "deserialize"]).create(ENode([], {}, ["Cannot deserialize ", obj$1.data, " because typeId \"", tid$3, "\" is not in the whitelist"]).toString());
        }m$21$0 = tid$3;if (m$21$0 === "native") {
          return obj$1.data;
        } else {
          m$22$0 = send(registry$0, tid$3);if (m$22$0 === void 0) {
            throw ErrorFactory(["kaiser", "deserialize"]).create(ENode([], {}, ["No deserializer for typeId \"", tid$3, "\""]).toString());
          } else {
            policy$6 = m$22$0;deser$0 = policy$6.deserialize((m$23$0 = (data$0 = obj$1.data, data$0), !recursive$1 ? object(items(data$0)) : (bridge$$1034$0 = m$23$0, (bridge$$1035$0 = bridge$$1034$0, (bridge$$1036$0 = bridge$$1035$0, (bridge$$1037$0 = bridge$$1036$0, bridge$$1037$0 === null || bridge$$1037$0 === void 0) || typeof bridge$$1036$0 === "string") || typeof bridge$$1035$0 === "number") || typeof bridge$$1034$0 === "boolean" ? data$0 : Array.isArray(m$23$0) ? (acc$9 = [], temp$13 = null, m$24 = null, (function () {
              $11: for (var _iterator3 = data$0[Symbol.iterator](), _step3; !(_step3 = _iterator3.next()).done;) {
                m$24 = _step3.value;
                var x$7 = undefined;x$7 = m$24;temp$13 = __at___$2.fromObject(x$7);acc$9.push(temp$13);
              }
            })(), acc$9) : getChecker(Object)(m$23$0) ? object((acc$10 = [], temp$14 = null, m$25 = null, (function () {
              $12: for (var _iterator3 = items(data$0)[Symbol.iterator](), _step3; !(_step3 = _iterator3.next()).done;) {
                m$25 = _step3.value;
                var k$2 = undefined;var v$2 = undefined;var t0$15 = undefined;var t1$6 = undefined;t0$15 = m$25;if (Array.isArray(t0$15) && (t1$6 = t0$15.length, t1$6 === 2)) {
                  k$2 = t0$15[0];v$2 = t0$15[1];temp$14 = [k$2, __at___$2.fromObject(v$2)];acc$10.push(temp$14);
                } else {
                  ___match_error(m$25);
                }
              }
            })(), acc$10)) : ___match_error(m$23$0, "Object? "))));
          }return deser$0;
        }
      }
    }
  } else {
    return ___match_error(m$18$0, "{match obj, recursive = true}");
  }
};Serializer$0.prototype.serialize = function serialize(obj$2) {
  var __at___$3 = undefined;var self$2 = undefined;__at___$3 = this;self$2 = this;return JSON.stringify(__at___$3.toObject(obj$2));
};Serializer$0.prototype.deserialize = function deserialize(repr$0) {
  var __at___$4 = undefined;var self$3 = undefined;__at___$4 = this;self$3 = this;return __at___$4.fromObject(JSON.parse(repr$0));
};Serializer$0.prototype.shallowCopy = function (obj$3) {
  var __at___$5 = undefined;var self$4 = undefined;__at___$5 = this;self$4 = this;return __at___$5.fromObject(__at___$5.toObject(obj$3, false), false);
};Serializer$0.prototype.deepCopy = function (obj$4) {
  var __at___$6 = undefined;var self$5 = undefined;__at___$6 = this;self$5 = this;return __at___$6.fromObject(__at___$6.toObject(obj$4));
};__amp____colon__(Serializer$0, __amp____colon__(($targ$11 = "Serializer", accum$5 = {}, accum$5["::name"] = $targ$11, accum$5), ($targ$12 = true, accum$6 = {}, accum$6["::egclass"] = $targ$12, accum$6)));default__minus__serializer$0 = Serializer$0();serialize$0 = default__minus__serializer$0.serialize.bind(default__minus__serializer$0);deserialize$0 = default__minus__serializer$0.deserialize.bind(default__minus__serializer$0);shallow__minus__copy$0 = default__minus__serializer$0.shallowCopy.bind(default__minus__serializer$0);deep__minus__copy$0 = default__minus__serializer$0.deepCopy.bind(default__minus__serializer$0);register$0(Object.prototype, __amp____colon__({ typeId: "builtin:Object", onDuplicate: "error", target: "direct" }, __amp____colon__((accum$7 = {}, accum$7.serialize = function serialize(obj$5) {
  return obj$5;
}, accum$7), (accum$8 = {}, accum$8.deserialize = function deserialize(obj$6) {
  return obj$6;
}, accum$8))));register$0(Date.prototype, __amp____colon__({ typeId: "builtin:Date", onDuplicate: "error", target: "direct" }, __amp____colon__((accum$9 = {}, accum$9.serialize = function serialize(obj$7) {
  return String(obj$7);
}, accum$9), (accum$10 = {}, accum$10.deserialize = function deserialize(s$0) {
  return new Date(s$0);
}, accum$10))));register$0(Error.prototype, __amp____colon__({ typeId: "builtin:Error", onDuplicate: "error", target: "direct" }, __amp____colon__((accum$11 = {}, accum$11.serialize = function serialize(obj$8) {
  var $targ$15 = undefined;var rval$1 = undefined;var $targ$13 = undefined;var $targ$14 = undefined;rval$1 = object(items(obj$8));$targ$13 = obj$8.message;rval$1.message = $targ$13;$targ$14 = obj$8.stack;rval$1.stack = $targ$14;if (send(obj$8, Symbol.errorTags)) {
    $targ$15 = send(obj$8, Symbol.errorTags);rval$1["::tags"] = $targ$15;void 0;
  }return rval$1;
}, accum$11), (accum$12 = {}, accum$12.deserialize = function deserialize(s$1) {
  var m$26 = undefined;var $targ$18 = undefined;var e$0 = undefined;var $targ$16 = undefined;e$0 = Error(s$1.message);m$26 = null;$13: for (var _iterator = items(s$1)[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
    m$26 = _step.value;
    var $targ$17 = undefined;var k$3 = undefined;var v$3 = undefined;var t0$16 = undefined;var t1$7 = undefined;t0$16 = m$26;if (Array.isArray(t0$16) && (t1$7 = t0$16.length, t1$7 === 2 && (k$3 = t0$16[0], v$3 = t0$16[1], !__in__(k$3, ["message", "::tags"])))) {
      $targ$17 = v$3;e$0[k$3] = $targ$17;void 0;
    } else {
      false;
    }
  }$targ$16 = s$1.stack;e$0.stack = $targ$16;if (s$1["::tags"]) {
    $targ$18 = s$1["::tags"];e$0[Symbol.errorTags] = $targ$18;void 0;
  }return e$0;
}, accum$12))));cmds$0 = { register: register$0, registerAll: register__minus__all$0, registerSingleton: register__minus__singleton$0, registerSingletons: register__minus__singletons$0, registerFunction: register__minus__function$0, registerFunctions: register__minus__functions$0 };m$27 = null;$0: for (var _iterator = (function () {
  var tmp$4 = undefined;return (tmp$4 = send(global, "@@KAISER", true), tmp$4 === undefined || tmp$4 === null ? tmp$4 : tmp$4.waitingList) || [];
})()[Symbol.iterator](), _step; !(_step = _iterator.next()).done;) {
  m$27 = _step.value;
  var cmd$0 = undefined;var args$0 = undefined;var t0$17 = undefined;var t1$8 = undefined;t0$17 = m$27;if (Array.isArray(t0$17) && (t1$8 = t0$17.length, t1$8 === 2)) {
    cmd$0 = t0$17[0];args$0 = t0$17[1];send(send(cmds$0, cmd$0, true), args$0);
  } else {
    ___match_error(m$27);
  }
}$targ$0 = module.exports;global["@@KAISER"] = $targ$0;$targ$19 = TYPEID$0;exports.TYPEID = $targ$19;$targ$20 = Serializer$0;exports.Serializer = $targ$20;$targ$21 = default__minus__serializer$0;exports.defaultSerializer = $targ$21;$targ$22 = serialize$0;exports.serialize = $targ$22;$targ$23 = deserialize$0;exports.deserialize = $targ$23;$targ$24 = shallow__minus__copy$0;exports.shallowCopy = $targ$24;$targ$25 = deep__minus__copy$0;exports.deepCopy = $targ$25;$targ$26 = register$0;exports.register = $targ$26;$targ$27 = register__minus__all$0;exports.registerAll = $targ$27;$targ$28 = register__minus__singleton$0;exports.registerSingleton = $targ$28;$targ$29 = register__minus__singletons$0;exports.registerSingletons = $targ$29;$targ$30 = register__minus__function$0;exports.registerFunction = $targ$30;$targ$31 = register__minus__functions$0;exports.registerFunctions = $targ$31;$targ$32 = true;exports.loaded = $targ$32;void 0;
